File: UIDescriptionWalker/UIDescriptionWalker/Roots.swift
//
//  Roots.swift
//  UIDescriptionWalker
//
//  Created by Damiaan on 31/07/2021.
//

extension BoardDescription {
	func getRoots() -> Set<String> {
		var roots = Set(genericDecomposition.keys)
		for (parent, children) in genericDecomposition {
			roots.subtract(children.filter{ $0 != parent })
		}
		return roots
	}
}

File: UIBoardViewer/src/lib/model/trees/roots.ts
import type {UIBoard} from '../BoardDescription'

export function getRoots(composition: UIBoard.GenericComposition): Set<string> {
	const dictionary = new Map(Object.entries(composition))
	const roots = new Set(dictionary.keys())

	for (const [parent, children] of dictionary) {
		for (const child of children) {
			if (child != parent) {
				roots.delete(child)
			}
		}
	}

	return roots
}